### PR TITLE
HPCC-30781 Introduce format registry for pluggable file formats

### PR DESCRIPTION
**Pull request by dcamper**
The goal of this PR is to define an easy-to-maintain interface between code that supports new file format types and the rest of the platform, at both compile and execution time.

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

The compiler successfully generates local a.out binaries (hthor format). It does crash when attempting to read Parquet files, as that code does not yet correctly extract the file name/path.

### CODE REVIEW

**Review by jackdelv — APPROVED** [review]
ParquetDiskRowReader changes look good. Other changes make sense as well.

**Review by ghalliday — COMMENTED** [review]
jlib_decl is the only change that is needed before merging

File: system/jlib/jfile.hpp [hunk 1]
```diff
@@ -788,4 +788,9 @@ extern jlib_decl IPropertyTreeIterator * getPlanesIterator(const char * category
 extern jlib_decl IFileIO *createBlockedIO(IFileIO *base, size32_t blockSize);
 extern jlib_decl size32_t getBlockedFileIOSize(const char *planeName, size32_t defaultSize=0);
 
+//---- Pluggable file type related functions ----------------------------------------------
+
+void addAvailableGenericFileTypeName(const char * name);
```
[the remark this answers is not in the document]
Comment (dcamper) [reply system/jlib/jfile.hpp:793]: Declarations added.

File: common/thorhelper/thorread.cpp [hunk 2]
```diff
@@ -1756,6 +1746,32 @@ IDiskRowReader * createDiskReader(const char * format, bool streamRemote, IDiskR
...
+    // at compile time
+    for (auto iter = genericFileTypeMap.begin(); iter != genericFileTypeMap.end(); iter++)
+        addAvailableGenericFileTypeName(iter->first.c_str());
+
+    return true;
+}
+
+MODULE_EXIT()
+{
+    genericFileTypeMap.clear();
```
[the remark this answers is not in the document]
Comment (dcamper) [reply common/thorhelper/thorread.cpp:1772]: Remove.

**Comment by dcamper** [comment]
@ghalliday I made some small changes based on your feedback.  Please give it a once-over.  Thanks!

**Review by ghalliday — APPROVED** [review]
Looks good.  Please can you squash the commits and I will merge

**Comment by dcamper** [comment]
@ghalliday Squashed and ready to merge.  Thanks!

**Pull request merged**